lib/vast/Dialect/HighLevel/HighLevelOps.cpp

```diff
@@ -273,42 +274,33 @@ namespace vast::hl
         st.addTypes(type);
     }
 
-    void CondOp::getSuccessorRegions(
-        llvm::Optional< unsigned > index,
-        llvm::ArrayRef< mlir::Attribute > operands,
-        llvm::SmallVectorImpl< mlir::RegionSuccessor > &regions
-    ) {
-        if (index) {
-            if (index.value() == 0) {
-                regions.push_back(trivial_region_succ(&getCondRegion()));
-            } else if (index.value() == 1) {
-                regions.push_back(trivial_region_succ(&getThenRegion()));
-                regions.push_back(trivial_region_succ(&getElseRegion()));
-            } else {
-                regions.push_back(mlir::RegionSuccessor(
-                    getOperation()->getResults())
-                );
-            }
-        } else {
-            regions.push_back(trivial_region_succ(&getCondRegion()));
-        }
-    }
-
-    bool CondOp::areTypesCompatible(mlir::Type lhs, mlir::Type rhs)
+    bool CondOp::typesMatch(mlir::Type lhs, mlir::Type rhs)
     {
         namespace tt = mlir::TypeTrait;
-        bool compatible = lhs == rhs
-            || lhs.hasTrait< tt::TypedefTrait >() || rhs.hasTrait< tt::TypedefTrait >()
-            || (lhs.hasTrait< tt::PointerTypeTrait >() && rhs.hasTrait< tt::PointerTypeTrait >());
-        VAST_ASSERT(compatible && "failed to verify CondOp types");
-        return compatible;
+        auto normalize = [](mlir::Type type) {
+           auto type_casted = type.dyn_cast< hl::LValueType >();
+           if (type_casted)
+               return type_casted.getElementType();
+            return type;
+        };
+        auto lhs_norm = normalize(lhs);
+        auto rhs_norm = normalize(rhs);
+        return lhs_norm == rhs_norm
+            || any_with_trait< tt::TypedefTrait >(lhs_norm, rhs_norm)
+            || all_with_trait< tt::PointerTypeTrait >(lhs_norm, rhs_norm);
     }
 
     logical_result CondOp::verifyRegions()
     {
         auto then_type = get_yielded_type(getThenRegion());
         auto else_type = get_yielded_type(getElseRegion());
-        return mlir::success(areTypesCompatible(then_type, else_type));
+        bool compatible = typesMatch(then_type, else_type);
+        if (!compatible)
+        {
+            llvm::errs() << "Failed to verify that return types " << then_type
```

Comment:
I do see a few times this condition hits and two region types does not match. This is what I see in the log if that helps:
```
Failed to verify that return types !hl.int, Failed to verify that return types !!hlhl..long< unsigned >, long< unsigned >! in CondOp regions match.
hl.int in CondOp regions match.
Failed to verify that return types !hl.int, !hl.long< unsigned > in CondOp regions match.
Failed to verify that return types !hl.long< unsigned >, !hl.int in CondOp regions match.
```

Reply:
@kumarak can you investigate this more, it seams like invalid AST to me. There need to be implicit integer cast from `hl.int` to `hl.long` somewhere.

Reply:
The error logs was due to not operation and got fixed with the PR #264.